pkg/controller/batchrelease/batchrelease_util.go

```diff
@@ -49,52 +54,55 @@ func initializeStatusIfNeeds(status *v1alpha1.BatchReleaseStatus) {
 	}
 }
 
-func signalStart(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.Phase == v1alpha1.RolloutPhaseHealthy {
-		return false
-	}
-	status.Phase = v1alpha1.RolloutPhaseHealthy
-	return true
-}
-
-func signalRestart(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.Phase == v1alpha1.RolloutPhaseInitial {
-		return false
-	}
+func signalRestart(status *v1alpha1.BatchReleaseStatus) {
 	resetStatus(status)
-	return true
 }
 
-func signalRecalculate(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.CanaryStatus.CurrentBatchState == v1alpha1.InitializeBatchState {
-		return false
-	}
+func signalReinitialize(status *v1alpha1.BatchReleaseStatus) {
 	status.CanaryStatus.CurrentBatchState = v1alpha1.InitializeBatchState
-	return true
 }
 
-func signalTerminating(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.Phase == v1alpha1.RolloutPhaseTerminating {
-		return false
-	}
+func signalStart(status *v1alpha1.BatchReleaseStatus) {
+	status.Phase = v1alpha1.RolloutPhaseHealthy
+	setCondition(status, v1alpha1.VerifyingBatchReleaseCondition, v1.ConditionTrue, "", "BatchRelease is verifying the workload")
+}
+
+func signalTerminating(status *v1alpha1.BatchReleaseStatus) {
 	status.Phase = v1alpha1.RolloutPhaseTerminating
-	return true
+	setCondition(status, v1alpha1.TerminatingBatchReleaseCondition, v1.ConditionTrue, "", "BatchRelease is terminating")
 }
 
-func signalFinalize(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.Phase == v1alpha1.RolloutPhaseFinalizing {
-		return false
-	}
+func signalFinalize(status *v1alpha1.BatchReleaseStatus) {
 	status.Phase = v1alpha1.RolloutPhaseFinalizing
-	return true
+	setCondition(status, v1alpha1.FinalizingBatchReleaseCondition, v1.ConditionTrue, "", "BatchRelease is finalizing")
 }
 
-func signalAbort(status *v1alpha1.BatchReleaseStatus) bool {
-	if status.Phase == v1alpha1.RolloutPhaseAbort {
-		return false
+func signalRecalculate(release *v1alpha1.BatchRelease, newStatus *v1alpha1.BatchReleaseStatus, info *workloads.WorkloadInfo) {
+	infoBy, _ := json.Marshal(info)
+	klog.Infof("workloadInfo: %v", string(infoBy))
+
+	// ensure current batch upper bound
+	currentBatch := int32(len(release.Spec.ReleasePlan.Batches) - 1)
+	if release.Spec.ReleasePlan.BatchPartition != nil {
+		currentBatch = integer.Int32Min(*release.Spec.ReleasePlan.BatchPartition, currentBatch)
 	}
-	status.Phase = v1alpha1.RolloutPhaseAbort
-	return true
+
+	// find the suitable current batch via ready replicas
+	if !util.IsControlledByRollout(release) && info != nil && info.Status != nil {
```

Comment:
why we need check whether release is controlled by rollout? it seems that rollout should depend upon the batchrelease, but batchrelease has no dependency with rollout